llvm/lib/CodeGen/PrologEpilogInserter.cpp

```diff
@@ -1164,6 +1170,7 @@ void PEI::calculateFrameObjectOffsets(MachineFunction &MF) {
 void PEI::insertPrologEpilogCode(MachineFunction &MF) {
   const TargetFrameLowering &TFI = *MF.getSubtarget().getFrameLowering();
 
+  TFI.emitCFIsForCSRsHandledByRA(MF, RDA);
```

Comment:
suppose we had the following code (before block layout is finalized):

```
ENTRY:
save x1
.cfi_offset x1, ...
...
beq ... %RET

BB:
...

RET:
ld x1, ...
.cfi_register x1, x1
...
ret
```

after the block layout the code may look like this (now `%RET` is layout successor of `%ENTRY`):


```
ENTRY:
save x1
.cfi_offset x1, ...
...
bne ... %BB

RET:
ld x1, ...
.cfi_register x1, x1

BB:
...
j %RET

...
ret
```

The cfi directives apply to all the code below it in the layout, so the cfi's are wrong now. The CFIInserter compares the CFI info from the layout successor to what it has to be and corrects it if necessary.

I basically copied that pass from the existing one `CodeGen/ CFIInstrInserter.cpp` and modified it.

Reply:
> I basically copied that pass from the existing one CodeGen/CFIInstrInserter.cpp and modified it.

Is it possible to modify `CodeGen/CFIInstrInserter.cpp` instead of having a RISC-V specific copy?

Reply:
Yes, it's possible. In some cases (notably when we have vector spills), we need to use cfi expression instead of simple `.cfi_offset`. I can add a small cfi-expression decoder to `CFIInserter` for that.

Here, I recorded cfi info into `RISCVMachineFunctionInfo` so it can be used by `RISCVCFIInserter`.